### ISSUE TITLE
About the current versioning policy with "next"

### ISSUE THREAD

**Issue by pat-s** [woodpecker-ci/woodpecker#1780]
### Clear and concise description of the problem

The current "next" version is in fact a new major release which is in "rolling" development for a long time now.

Reading the roadmap in #869 let's one assume that there won't be an "official" release soon assuming all points listed there must be met.

This fact combined with the unclear state on this topic gives the project a somewhat "unprofessional" touch and also comes with versioning issues for admins. I.e. admins must work with sha checkouts based on commits and it is hard to get an easy changelog about recent changes.

Additional confusion comes in with the "old" 0.15.x releases which let many people start with this version first. However, people often face a lot of missing features and they get redirected to "next" then. Yet to upgrade they need to change many things, both users and admins, to compl

Given that so many people are on "next" meanwhile and "next" seems to be usable for many - agnostic from the backend - I would like to discuss whether the versioning approach could be rethought.
I am well aware that not everything is working perfectly yet but I don't think that users always expect an 1.0.0 version to be "fully stable" and "finished" in all areas.

### Suggested solution

Release a 1.0.0 version and continue with a semantic versioning release policy for upcoming changes. This makes it a lot easier to follow development for admins and allows to have a "clean" changelog. Especially changes like #1765 which might break usability for specific OS might get unnoticed in the current versioning approach.

### Validations

- [X] Checked that the feature isn't part of the `next` version already [https://woodpecker-ci.org/faq#which-version-of-woodpecker-should-i-use]
- [X] Read the [Contributing Guidelines](https://github.com/woodpecker-ci/woodpecker/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't already an [issue](https://github.com/woodpecker-ci/woodpecker/issues) that request the same feature to avoid creating a duplicate.

**Comment by pat-s** [comment]
Any comment/thought from the maintainers @anbraten @6543?

**Comment by lafriks** [comment]
Imho we should finish with #567 before 1.0 to have single breaking change in pipeline yaml format

**Comment by 6543** [comment]
Well the issue list is getting smaler and i'm currently focusted on bugfixing plus clean up low hanging fruites that do block 1.0.0

Beside the group vs depend-on dessision for steps  ... i would say all big breaking changes are mostly done

**Comment by 6543** [comment]
And without a good reason i wont add anythin to the baglock ... more i move things into v1.1.0 if they are non breaking & blocking

**Comment by nitram509** [comment]
> Well the issue list is getting smaler and i'm currently focusted on bugfixing plus clean up low hanging fruites that do block 1.0.0
> 
> Beside the group vs depend-on dessision for steps ... i would say all big breaking changes are mostly done

That sounds great.
is there a (check)list, of things to do, that others maybe could support?

**Comment by zc-devs** [comment]
Waterfall vs Agile 😄 

@pat-s, @nitram509, 99% agree with you.

> Release a 1.0.0 version and continue with a semantic versioning release policy for upcoming changes

Suppose current `next` is released. According to semver it is MAJOR and should be 1.x.x. There are a lot of breaking changes that cannot be released "today". So,  if we release more often, then there will be a lot of major versions. What's about support? It is hard work to support a few versions. And I understand desire of maintainers to include all breaking changes in one release.

What if we leave semantic versioning as currently is, but introduce calendar versioning for the `next` branch (only)? Look at MinIO ([Releases](https://github.com/minio/minio/releases), [Release Policy](https://github.com/minio/minio/discussions/13737)). I think it addresses issues and suggestions mentioned above, but also eliminates support burden for maintenance each major version.

**Comment by 6543** [comment]
We already tag next-commitid based to be able to pin a specific next version ...

Open issues for v1.0.0 are now down to 27 ... so it never looked that promising ;)
And pulls to adress the hardest open issues are open for review or at least wip

I encourage all here to just hve a look at the issues labeld with docs ... that will brind down the number and is easy to address

**Comment by pat-s** [comment]
> Suppose current next is released. According to semver it is MAJOR and should be 1.x.x. There are a lot of breaking changes that cannot be released "today". So, if we release more often, then there will be a lot of major versions. What's about support? It is hard work to support a few versions. And I understand desire of maintainers to include all breaking changes in one release.

That's just how the game works. Major releases are (in my vision) just an indicator for breaking changes, i.e. action is required by the users. It just says "read the changelog carefully and don't blindly update". Yes, one certainly does not want to have such release every few weeks - but this is also not what I am aiming for. There's a sweet spot between not releasing a "usable" version for over a year and the former.

I don't mind a software running on 15.12.x - as long as the release and changelog make sense and do "their job". 
AFAIU WP and others are not aiming at backporting and supporting dozens of "major" versions at the same time. If this would be the case, things would be a bit different of course.

Also I am fully aware that the above is slightly subjective - though I would also argue it's the implicit default which most people agreed on in (public) software development.

> And I understand desire of maintainers to include all breaking changes in one release.

Being a maintainer myself and applying semver, I don't. With my main point being that it's hard to follow the changes in `next` (which is de-facto the release most people use or *should* be using, not only because the 0.15.x hasn't seen any improvement since the dev time of `next`). I am aware of the images tags by commit and PR-based images (which both is great) but I would like to see a curate changelog from time to time that describes the changes on a higher-level, possibly with some examples and background information rather than having to go through the commit history and figure changes and their reasonings out myself.

I am aware that I can't change this and that the current philosophy is different here but I can at least bring the topic up for discussion and express my thoughts.

> Open issues for v1.0.0 are now down to 27 ... so it never looked that promising ;)

I appreciate everyone's work but 27 doesn't sound like a "small" number 🙃 And phrased in another way: most of these are probably "not breaking" and more like "small enhancements"? (haven't gone through the list though).

Also I hope that after 1.0.0, whenever it will happen, there won't be another "next" but maybe a "normal" release cycle of a few months with a curated changelog 💟 . 

Last I'd like to express my gratitude with the project, I like it a lot (the idea, the motivation, the product) and I/we are also about to start sponsoring it ❤️ . I don't want to sound mainly negative in this issue, my goal is just to bring up the discussion and maybe start a thought process through it *to make something better*.

**Comment by 6543** [comment]
Well what we could do is to decide on some guidelines like if we hit max 3 breaking changes and/or 5features ... a release should be prepaired

**Comment by runephilosof-karnovgroup** [comment]
> It is hard work to support a few versions

You don't need to support old major versions. Semver does not require that.
It will be an improvement to release 1.0.0 now and just up the major version every time a breaking change is released, without maintaining old major releases.

**Comment by mvdkleijn** [comment]
First of all, I also think this is a great project with a lot of potential and you have my many thanks for its existence. :+1: 

**TLDR;** I agree with the sentiments of @pat-s @nitram509 and @zc-devs they shared above with regards to using semver. Breaking changes in a non-major version is fine _if_ the major version is zero.

I'd like to share some experiences and thoughts, my apologies for being long-winded:

**First use experience**

Some time ago, I started using Woodpecker instead of Drone. Fair warning, I'm gonna be brutally honest here.

The experience was... horrible. Sorry, but it was. Not trying to be negative here but I got to be honest for clarity's sake. :smiley: 

Setting up Drone cost me about one hour with a hello world pipeline building successfully. When I learned about Woodpecker CI, I wanted to support the pure FOSS project and switched. Looking at the site and docs, I started off with the `v0.15.x` version as that seemed to be the "stable" version.

I spent quite some time trying to get everything setup. Docs were unclear, communication between server and agent kept breaking for unknown reasons & with unclear error messages, pipelines were unreliable. I looked through the docs, the Github issues, outstanding PRs and even the code for hints to my problems.

**Solution**

Out of sheer desperation I tried the "next" version which I would normally not do since I don't run development branches in production normally.

Though I still had a number of issues, mainly in a lack of good quick start documentation, stuff magically started working. On top of that, I suddenly had a bunch of features that I missed. It still took way more time to setup properly than with Drone, but if I'd started off with "next" instead of the `v0.15.x` version, I'd have saved myself quite some time and frustration.

**My conclusions / suggestions:**
- Consider `v0.15.x` legacy & stop supporting it
- Release a `v0.50.0` version with the current content of "next"
- Apply breaking changes to the `v0.50.x+` versions
- In the site & docs, clearly point people to `v0.50.x+` as the place to start
- Start a clear changelog from `v0.50.x` onwards (i.e. no need to backfill)
- When you feel like breaking changes are more uncommon, graduate to `v1.0.0`

**On semver**

In semver, the MAJOR version number does indeed signify breaking changes. However, the zero (0) major version is special. It signifies that _"we're in active development towards a `v1.0.0` and stuff should be expected to break"_. (see: https://semver.org/#spec-item-4 & https://semver.org/#how-should-i-deal-with-revisions-in-the-0yz-initial-development-phase)

In other words, breaking changes before you release a `v1.0.0` are fine in a minor version.

Like @runephilosof-karnovgroup stated, you don't need to support older versions. It is up to the admin to keep up-to-date with releases. If there's a bug in an older version, you fix it in a newer version. That's the whole point.

Again, my apologies for the long-winded post here... it comes from good intentions. :smile:

**Comment by zc-devs** [comment]
In regards to support... Imagine...

You are on v1.11.2 (latest v1). Yesterday version 2.0.0 is released with breaking changes, changelog and all this stuff. Today some critical CVE has been discovered and fixed/released in v2.0.1 and won't be backported to v1 (right?).

How do you feell? What are you going to do?

*you - is not particular person, but who says `we don't need support for previous version(s)`
*though, obviously, product owners may decide to go this way in their release policy...

And as of today developers still support previous release (even though it's useless in my opinion) ❤️ Hope, they will in the future after v1.

**Comment by 6543** [comment]
CVEs are tryed to be backported at least one version ... so to say v0.15 would get it in this context

**Comment by mvdkleijn** [comment]
My comment with regards to supporting previous releases was mostly geared towards supporting 2,3 or even more older releases. Supporting latest-1 is fine by me.

**Comment by nitram509** [comment]
> In regards to support... Imagine...
> 
> You are on v1.11.2 (latest v1). Yesterday version 2.0.0 is released with breaking changes, changelog and all this stuff. Today some critical CVE has been discovered and fixed/released in v2.0.1 and won't be backported to v1 (right?).
> 
> How do you feell? What are you going to do?
> 
> *you - is not particular person, but who says `we don't need support for previous version(s)` *though, obviously, product owners may decide to go this way in their release policy...
> 
> And as of today developers still support previous release (even though it's useless in my opinion) ❤️ Hope, they will in the future after v1.

Well, if I imagine being on Woodpecker v1, I would obviously hope for getting this fix backported.

That said, the world is not black and white. I mean setting a clear message/guideline, that older versions are not further supported, could still be the easier choice for the maintainers of Woodpecker. On-Top, on a "per use case", a really critical CVE (thinking of ssl heartbleed, or log4j, etc.) still can be an exception of the mentioned rule and thus backported.
You could even formalize that: e.g. "IF an issue for backporting a critical CVE fix has > 50 (or any other number) thumbs up, then it will be ...".
Such guidelines help manage expectations and help protect the volunteering contributors of this project.

Just my 2cent ;)

**Comment by runephilosof-karnovgroup** [comment]
@zc-devs wrote:
> You are on v1.11.2 (latest v1). Yesterday version 2.0.0 is released with breaking changes, changelog and all this stuff. Today some critical CVE has been discovered and fixed/released in v2.0.1 and won't be backported to v1 (right?).

Suppose being on b1787f8 (next) and then https://github.com/woodpecker-ci/woodpecker/issues/1743 is merged containing a breaking change.
Today some critical CVE was discovered and fixed/released in 7f725a5, no backport.
You update, having no way to notice the breaking change before doing so.
Result, your system is down.

Even with no backporting, I still think it is an improvement over the current state.
But as @6543 wrote:
> CVEs are tryed to be backported at least one version ... so to say v0.15 would get it in this context

So v0.15 already being treated as v1.x. Releasing next as v2.x, would mean that the next breaking change would make v1.x unsupported. Luckily, this is an open source project, so if the community backports an issue to an older version, there is a somewhat high probability that it will get released after all.

@mvdkleijn wrote:
> However, the zero (0) major version is special. It signifies that "we're in active development towards a v1.0.0 and stuff should be expected to break".

They also write https://semver.org/#how-do-i-know-when-to-release-100
> If your software is being used in production, it should probably already be 1.0.0. If you have a stable API on which users have come to depend, you should be 1.0.0. If you’re worrying a lot about backward compatibility, you should probably already be 1.0.0.

**Comment by pat-s** [comment]
This discussion got a bit off-topic and more into "backporting and supporting multiple versions". My initial proposal was mainly focused on the long developing time of "next"/1.0.0 and that, at some point, everything was just added to "next" without a clear idea when there will be release or a "stop" for new features.

Not sure if this issue helped but now there is a shortlisted milestone and already a changelog documentation (WIP) and some features have been moved to a 1.1.0 milestone. For "professional" use this is a good thing and the main point I wanted to address :)

Looking forward having a "proper" 1.0.0 release soon followed by "normal" feature releases coming in reasonable time frames :) Thanks @anbraten @6543 and everybody else doing great work.

**Comment by 6543** [comment]
to @runephilosof-karnovgroup 's backport concerns ...

as long as somebody do provide a working patch we can release I dont mind to let it be released even for our v0.14 branch ... it's just that this is not the woodpecker-maintainers response in that regard ...

**Comment by 6543** [comment]
... as long as the pipeline to release things does work or can be fixed easily for that branch

**Comment by zc-devs** [comment]
> And I understand desire of maintainers to include all breaking changes in one release.

When I said that I thought about backend for mobile applications. But I've found [beautiful example here](https://github.com/woodpecker-ci/woodpecker/issues/267). Support is going at next level ;)

**Comment by anbraten** [comment]
As written before we all agree on this. Pleas hold your horses a bit. We will release `v1.0` asap mainly waiting for #1743 to be merged

**Comment by anbraten** [comment]
1.0 got release and the idea is to release more often now if there are new features / fixes.
